subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/DiscoverInputs.java

```diff
@@ -0,0 +1,203 @@
+        return ImmutableSortedSet.copyOf(discoveredInputs);
+    }
+
+    private void addDiscoveredInputsToTask(IncrementalTaskInputsInternal inputs, ImmutableSortedSet<File> discoveredInputs) {
+        for (File discoveredInput : discoveredInputs) {
+            inputs.newInput(discoveredInput);
+        }
+    }
+
+    private void writeDiscoveredInputsToFile(ImmutableSortedSet<File> discoveredInputs) throws IOException {
+        File outputFile = getDiscoveredInputs().getAsFile().get();
+        final BufferedWriter outputStreamWriter = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(outputFile), Charsets.UTF_8));
+        try {
+            for (File discoveredInput : discoveredInputs) {
+                outputStreamWriter.write(discoveredInput.getAbsolutePath());
+                outputStreamWriter.newLine();
+            }
+        } finally {
+            IOUtils.closeQuietly(outputStreamWriter);
+        }
+    }
+
+    private IncrementalCompileProcessor createIncrementalCompileProcessor(List<File> includeRoots) {
+        PersistentStateCache<CompilationState> compileStateCache = compilationStateCacheFactory.create(getPath());
+        DefaultSourceIncludesParser sourceIncludesParser = new DefaultSourceIncludesParser(sourceParser, importsAreIncludes.getOrElse(false));
+        DefaultSourceIncludesResolver dependencyParser = new DefaultSourceIncludesResolver(includeRoots);
+        IncrementalCompileFilesFactory incrementalCompileFilesFactory = new IncrementalCompileFilesFactory(sourceIncludesParser, dependencyParser, hasher);
+        return new IncrementalCompileProcessor(compileStateCache, incrementalCompileFilesFactory);
+    }
+
+    @Input
+    protected Collection<String> getIncludePaths() {
+        if (includePaths == null) {
+            Set<File> roots = includes.getFiles();
+            ImmutableList.Builder<String> builder = ImmutableList.builder();
+            for (File root : roots) {
+                builder.add(root.getAbsolutePath());
+            }
+            includePaths = builder.build();
+        }
+        return includePaths;
+    }
+
+    /**
+     * Add directories where the compiler should search for header files.
+     */
+    public void includes(Object includeRoots) {
+        includes.from(includeRoots);
+    }
+
+    /**
+     * Returns the source files to be compiled.
+     */
+    @InputFiles
+    @SkipWhenEmpty
+    public ConfigurableFileCollection getSource() {
+        return source;
+    }
+
+    /**
+     * Adds a set of source files to be compiled. The provided sourceFiles object is evaluated as per {@link org.gradle.api.Project#files(Object...)}.
+     */
+    public void source(Object sourceFiles) {
+        source.from(sourceFiles);
+    }
+
+    @OutputFile
+    public RegularFileVar getDiscoveredInputs() {
+        return discoveredInputs;
+    }
+
+    @Input
+    public PropertyState<Boolean> getImportsAreIncludes() {
+        return importsAreIncludes;
+    }
```

Comment:
Is your question is rhetorical? We could provide a setter for various "things" which return a `boolean`, like `boolean`, `Provider<Boolean>`, `Callable<Boolean>`, etc. This is what we would need to do for every property of `PropertyState` out there. Instead, `PropertyState` provides some `set` methods for all those things. So I only need to expose this getter and the client can the call the `set` for whatever suits him best + I don't need to implement overloads for the setters + if we add some more convenience setters to `PropertyState` my current property supports those automatically.

Reply:
It wasn't a rhetorical question. I understand why we think it's great to have everything in one place as the _implemetor._ But as a _user,_ for the typical use case (especially in the case of a simple `boolean` parameter) this is much better:

```groovy
task.importsAreIncludes = true
```

Than this:

```groovy
task.importsAreIncludes.set(true)
```

The difference in Java is even more pronounced, compare:

```java
task.setImportsAreIncludes(true);
```

vs.

```java
task.getImportsAreIncludes().set(true);
```

I think it's great that we can now support all kinds of complex ways to set the property value via the `PropertyState` when needed. But 95% of the time people will just want to set a property to a value, and we are making that more complicated here.

I think it would be much better if we could set it as a simple `boolean`, too.

Reply:
FTR: The only use case I have is to set a `Provider<Boolean>`.